### PR TITLE
Add sorted builtin support for autograph

### PR DESCRIPTION
**Pull request by neyugncol**
This PR adds `sorted` builtin support for tensor in autograph.
Since `sorted` has useful sort by comparison key, it might make sense to add `sorted`
as the supported list of ops for autograph.

### CODE REVIEW

File: tensorflow/python/autograph/operators/py_builtins.py [hunk 1]
```diff
@@ -461,8 +464,41 @@ def _py_all(iterable):
...
+
+def _tf_sorted(iterable, key, reverse):
+  """Overload of sorted_ for Tensor iterable."""
+  if reverse is UNSPECIFIED:
+    direction = 'ASCENDING'
+  else:
+    direction = 'DESCENDING'
+  if key is not UNSPECIFIED:
+    mapped = parallel_ops.vectorized_map(key, iterable)
+    with ops.control_dependencies(
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/py_builtins.py]: Optional: If the rank is static, you could add an extra check for an early warning:

```
if mapped.shape.rank is not None and mapped.shape.rank != 1:
  raise ValueError('sort only supports only 1D tensors')
```

**Review by mdanatg — DISMISSED** [review]
Nice!

File: tensorflow/python/autograph/operators/py_builtins.py (hunk 1 above)
Comment (lyonguyen8697) [reply tensorflow/python/autograph/operators/py_builtins.py]: Thanks @mdanatg, the PR has been updated.

**Review by mdanatg — DISMISSED** [review]

**Comment by gbaned** [comment]
@lyonguyen8697 Can you please fix the build failures? Thanks!

**Review by mdanatg — DISMISSED** [review]

**Review by mdanatg — DISMISSED** [review]
Note that you can save a few roundtrips by testing locally. To do that, you can follow the instructions at https://www.tensorflow.org/install/source and then run `bazel test //tensorflow/python/autograph/...`

**Comment by gbaned** [comment]
@lyonguyen8697 Can you please fix the build failures? Thanks!

**Comment by mdanatg** [comment]
I think this will be tricky to address since there's a circular dependency, possibly by parallel_for. We're working on a broader effort to remove these circular deps, but in the mean time, I think this will need a workaround:
 * remove the parallel_for dep from the BUILD file (if tests fail, add the dependency to the tests alone) - bazel rejects circular imports outright
 * in the Pyhton file, use LazyLoader to import the module (look at the existing one for an example) - that will prevent import errors due to these cycles

**Comment by lyonguyen8697** [comment]
@mdanatg thanks for your help. So I removed `parallel_for` from BUILD file and added this code to python file:
`parallel_ops = lazy_loader.LazyLoader(
    'parallel_ops', globals(),
     'tensorflow.python.ops.parallel_for.control_flow_ops'
)`
Am I doing this correctly?

**Comment by lyonguyen8697** [comment]
It's still raise an import error. Look like the recently added `gen_spectral_ops` in 3a552b92f4ab783ca9984fe9a350aafec1e32349 that cause the error.
`Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/autograph/operators/py_builtins_test.py", line 395, in test_sorted_tensor
    py_builtins.sorted_(iterable_1, key=lambda x: -x))), [3, 2, 1])
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/autograph/operators/py_builtins.py", line 472, in sorted_
    return _tf_sorted(iterable, key, reverse)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/autograph/operators/py_builtins.py", line 483, in _tf_sorted
    mapped = parallel_ops.vectorized_map(key, iterable)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
    module = self._load()
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
    module = importlib.import_module(self.__name__)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/ops/parallel_for/control_flow_ops.py", line 34, in <module>
    from tensorflow.python.ops.parallel_for.pfor import PFor
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/ops/parallel_for/pfor.py", line 53, in <module>
    from tensorflow.python.ops import gen_spectral_ops
ImportError: cannot import name 'gen_spectral_ops'`

**Comment by mdanatg** [comment]
Ok, that was somewhat expected. This is why we need to fix these terrible circular dependencies. Anyhow, this should be resolved by adding the pfor dependency (the one you had to remove from the BUILD file) to the test target only, that is to `autograph/operators/BUILD`, under the rule with `name = "py_builtins_test"`.

File: tensorflow/python/autograph/operators/BUILD [hunk 2]
```diff
@@ -114,6 +114,7 @@ py_test(
         "//tensorflow/python/autograph/core",
         # TODO(b/145618471): Remove this transitive dependency.
         "//tensorflow/python/distribute:input_lib",
+        "//tensorflow/python/ops/parallel_for:control_flow_ops",
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/BUILD]: Can you try to also include `//third_party/tensorflow/python/ops/signal`?

**Comment by mdanatg** [comment]
@allenlavoie do you know which dependency is still missing here? Pulling `//tensorflow/python/ops/parallel_for:control_flow_ops` alone gives an import error inside pfor.py:

```
opt/bin/tensorflow/python/autograph/operators/py_builtins_test.runfiles/org_tensorflow/tensorflow/python/ops/parallel_for/pfor.py", line 53, in <module>
    from tensorflow.python.ops import gen_spectral_ops
ImportError: cannot import name 'gen_spectral_ops'
```

File: tensorflow/python/autograph/operators/BUILD (hunk 2 above)
Comment (mdanatg) [reply tensorflow/python/autograph/operators/BUILD]: Also, please be sure to indent consistently, otherwise the lint checks will fail.

**Comment by allenlavoie** [comment]
Probably //tensorflow/python/ops/parallel_for:parallel_for makes more sense there? But @agarwal-ashish would know for sure.

File: tensorflow/python/autograph/operators/BUILD (hunk 2 above)
Comment (lyonguyen8697) [reply tensorflow/python/autograph/operators/BUILD]: Thanks @mdanatg , the PR has been updated.

**Review by mdanatg — APPROVED** [review]

**Pull request merged**